src/Http/Routing/src/RouteEndpointDataSource.cs

```diff
@@ -0,0 +1,264 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection;
+using System.Runtime.CompilerServices;
+using Microsoft.AspNetCore.Builder;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Patterns;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.Extensions.FileProviders;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNetCore.Routing;
+
+internal sealed class RouteEndpointDataSource : EndpointDataSource
+{
+    private readonly List<RouteEntry> _routeEntries = new();
+    private readonly IServiceProvider _applicationServices;
+    private readonly bool _throwOnBadRequest;
+
+    public RouteEndpointDataSource(IServiceProvider applicationServices, bool throwOnBadRequest)
+    {
+        _applicationServices = applicationServices;
+        _throwOnBadRequest = throwOnBadRequest;
+    }
+
+    public ICollection<Action<EndpointBuilder>> AddEndpoint(
+        RoutePattern pattern,
+        Delegate routeHandler,
+        IEnumerable<string>? httpMethods,
+        bool isFallback)
+    {
+        RouteEntry entry = new()
+        {
+            RoutePattern = pattern,
+            RouteHandler = routeHandler,
+            HttpMethods = httpMethods,
+            IsFallback = isFallback,
+            Conventions = new ThrowOnAddAfterBuildCollection(),
+        };
+
+        _routeEntries.Add(entry);
+
+        return entry.Conventions;
+    }
+
+    public override IReadOnlyList<Endpoint> Endpoints
+    {
+        get
+        {
+            var endpoints = new List<RouteEndpoint>(_routeEntries.Count);
+            foreach (var entry in _routeEntries)
+            {
+                endpoints.Add((RouteEndpoint)CreateRouteEndpointBuilder(entry).Build());
+            }
+            return endpoints;
+        }
+    }
+
+    // I just want to point out that we're using fancy covariant returns introduced by C# 9! Not that it matters here :D
+    public override IReadOnlyList<RouteEndpoint> GetEndpointGroup(RouteGroupContext context)
+    {
+        var endpoints = new List<RouteEndpoint>(_routeEntries.Count);
+        foreach (var entry in _routeEntries)
+        {
+            endpoints.Add((RouteEndpoint)CreateRouteEndpointBuilder(entry, context.Prefix, context.Conventions).Build());
+        }
+        return endpoints;
+    }
+
+    public override IChangeToken GetChangeToken() => NullChangeToken.Singleton;
+
+    // For testing
+    internal RouteEndpointBuilder GetSingleRouteEndpointBuilder()
+    {
+        if (_routeEntries.Count != 1)
+        {
+            throw new InvalidOperationException($"There are {_routeEntries.Count} endpoints defined! This can only be called for a single endpoint.");
```

Comment:
I didn't see a test for this error case. Is it because it's not really possible today?

Reply:
This really only called by our tests. I put this `throw` here only to make things easier on ourselves later if we mess up authoring a test and define multiple endpoints but expect a single builder..

There are a bunch of existing tests that define a single endpoint using `MapGet` or similar and want to get a single `RouteEndpointBuilder`. They all only define one endpoint, because defining more would just make it harder to get the right builder.